File: 03_query_ai/02_ollama.R
# workflow.R

# A script to query the lightweight LLM gemma-3-1b!

# Starting message
cat("\n🚀 Sending LLM query in R...\n")

# If you haven't already, install these packages...
# install.packages(c("httr2", "jsonlite"))

# Load libraries
library(httr2)    # For HTTP requests
library(jsonlite) # For working with JSON

PORT = 11434
OLLAMA_HOST = paste0("http://localhost:", PORT)
url = paste0("http://localhost:", PORT, "/api/generate")

# Construct the request body as a list
body <- list(
  model = "smollm2:1.7b", # Model name
  prompt = "Is model working?",        # User prompt
  stream = FALSE       # Non-streaming response
)

# Build and send the POST request to the LM Studio REST API
res <- request(url) %>%
  req_body_json(body) %>%   # Attach the JSON body
  req_method("POST") %>%   # Set HTTP method
  req_perform()             # Execute the request

# Parse the response JSON
response <- resp_body_json(res)

# Extract the model's reply as a character string
output = response$choices[[1]]$message$content

# Print the model's reply
cat(output)

# Closing message
cat("\n✅ LLM query complete. Exiting R...\n")

# Close out of R, and don't save the environment.
q(save = "no")

File: 03_query_ai/02_ollama.py
# 02_ollama.py
# Query Ollama LLM
# Pairs with 02_ollama.R
# Tim Fraser

# This script demonstrates how to query a local Ollama LLM instance using Python.
# Students will learn how to make HTTP POST requests to interact with language models
# running locally on their machine.

# 0. SETUP ###################################

## 0.1 Load Packages #################################

import requests  # for HTTP requests
import json      # for working with JSON

# Starting message
print("\n🚀 Sending LLM query in Python...\n")

# If you haven't already, install the requests package...
# pip install requests

## 0.2 Configure Connection #########################

# Set the port where Ollama is running
PORT = 11434
OLLAMA_HOST = f"http://localhost:{PORT}"
url = f"{OLLAMA_HOST}/api/generate"

## 0.3 Construct Request Body #######################

# Build the request body as a dictionary
# This tells Ollama which model to use and what prompt to send
body = {
    "model": "smollm2:1.7b",  # Model name
    "prompt": "Is model working?",  # User prompt
    "stream": False  # Non-streaming response
}

# 1. SEND REQUEST ###################################

# Build and send the POST request to the Ollama REST API
# The requests library makes it easy to send HTTP requests
response = requests.post(url, json=body)

# 2. PARSE RESPONSE ################################

# Parse the response JSON
# The response from Ollama is in JSON format
response_data = response.json()

# Extract the model's reply as a string
# The response structure contains the generated text
output = response_data["response"]

# Print the model's reply
print(output)

# Closing message
print("\n✅ LLM query complete. Exiting Python...\n")
